File: solutions/PHP/smallestNumber.php
<?php

/****
PHP class to find the smallest element in a rotated sorted array.

Author: Uthman Dev

Description: This class uses binary search to find the minimum element in
             a rotated sorted array, where the array was originally sorted in
             ascending order and then rotated.
****/

class RotatedArrayMinFinder
{
    private array $arr;

    // Constructor to initialize the array
    public function __construct(array $arr)
    {
        $this->arr = $arr;
    }

    // Method to find the smallest element in the rotated sorted array
    public function findSmallest(): int
    {
        $left = 0;
        $right = count($this->arr) - 1;

        // If the array is not rotated, return the first element
        if ($this->arr[$left] < $this->arr[$right]) {
            return $this->arr[$left];
        }

        // Binary search to find the minimum element
        while ($left < $right) {
            $mid = floor(($left + $right) / 2);

            // If mid element is greater than the rightmost element,
            // the minimum is in the right half
            if ($this->arr[$mid] > $this->arr[$right]) {
                $left = $mid + 1;
            }
            // Otherwise, the minimum is in the left half (including mid)
            else {
                $right = $mid;
            }
        }

        // After the loop, left will point to the minimum element
        return $this->arr[$left];
    }
}

// Example usage:
$arr = [4, 5, 6, 7, 0, 1, 2];
$finder = new RotatedArrayMinFinder($arr);
echo "The minimum value is: " . $finder->findSmallest(); // Output: 0

?>

File: solutions/PYTHON/smallestNumber.py
# Python function to find the minimum element in a rotated sorted array.
#
# Author: Uthman Dev
#
# Description: This function takes a rotated sorted array and uses binary search to
#              find the minimum element in the array. If the array is not rotated, 
#              it returns the first element as the minimum.
#
# Parameters:
# arr (list): A rotated sorted list of integers.
#
# Returns:
# int: The minimum element in the array.

def find_minimum_in_rotated_array(arr):
    left = 0
    right = len(arr) - 1

    # Special case: If the array is not rotated
    if arr[left] < arr[right]:
        return arr[left]

    # Binary search to find the minimum element
    while left < right:
        mid = (left + right) // 2

        # Debugging output to verify pointer positions
        print(f"Left: {left}, Right: {right}, Mid: {mid}, Mid Element: {arr[mid]}")

        # If mid element is greater than the rightmost element, the minimum is in the right half
        if arr[mid] > arr[right]:
            left = mid + 1
        # Otherwise, the minimum is in the left half or at mid
        else:
            right = mid

    # At the end of the loop, left will be the index of the minimum element
    return arr[left]

# Example usage
arr = [4, 5, 6, 7, 0, 1, 2]  # Rotated sorted array
result = find_minimum_in_rotated_array(arr)  # Call the function with the array

# Print the result
print("The minimum value is:", result)  # Expected output: 0